### ISSUE TITLE
Why does column take precedence over an object with the same name in data table in cases of join with a subsetted data table

### ISSUE THREAD

**Issue by wsrfad** [Rdatatable/data.table#4946]
Both `operator_descriptions[condition]` and `operator_conditions[operator_descriptions, on="operator"]` work without issue, yet (as you demonstrate) the merge fails.

`
operator_descriptions <- data.table(operator = "/", description = "/")
  operator_conditions <- data.table(operator = "/", condition = "Quotient non zero")
  condition <- c(TRUE)
  operator_conditions[operator_descriptions[condition], on = c("operator")]`

`Output
Error: When i is a data.table (or character vector), the columns to join by must be specified either using 'on=' argument (see ?data.table) or by keying x (i.e. sorted, and, marked as sorted, see ?setkey). Keyed joins might have further speed benefits on very large data due to x being sorted in RAM.`

Despite the fact that condition is in the scope of operator_descriptions, condition column of operator_conditions seems to take precedence over the condition object defined in previous statement.

This behavior requires one to be cautious about names etc. While in normal cases, this is fine as we intend to use column names mostly. But in cases like these, I am not sure if it is the intention.

`#` `Output of sessionInfo()`
`R version 3.4.4 (2018-03-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.7 LTS`

**Comment by avimallu** [comment]
I am not sure of what the exact content is of your post, but I presume that it's got to do with overlapping names in different scopes - one as a column name inside a `data.table`, and the other as an object in R's global (or other) environment. I would point you to [Jan's comment](https://github.com/Rdatatable/data.table/issues/4878#issuecomment-762726361), on a similar issue. 

An excerpt from that comment is: _it is fragile to rely on a behavior in such situations_, which this situation will be a subset of. `data.table` cannot be sure of the user's intent behind using a conflicting reference - and by default/design, chooses to prefer a column defined inside it with the same name over another object in R's environment. 

TL;DR: Don't use overlapping names in varying scopes for now (it's also _probably_ not best practice imho), or wait until #4304 is implemented.

**Comment by cdriveraus** [comment]
Just ended up here after posting https://stackoverflow.com/questions/67728442/r-data-table-scoping-reliably-refer-to-unknown-column-name-using-variable
Happy to hear a fix is on the way, just wanted to point out that it's not so straightforward to avoid overlapping names if one allows user specified names in a function...
